# Smester Project: Movie Recommendation System

## About Dataset
- This project uses a movie dataset collected from The Movie Database (TMDb), containing detailed information about thousands of films. The dataset is designed for data analysis, visualization, and machine learning applications related to the film industry.
- The dataset includes important movie attributes such as:
     - Movie titles and overviews
     - Cast and crew information
     - Genres and keywords
     - Budget and revenue
     - Release dates
     - Production companies and countries
     - Popularity scores and audience ratings
     - Runtime, language, and movie status

- A key advantage of this dataset is the availability of complete cast and crew records, along with updated revenue information. The data enables analysis of factors influencing movie success, audience ratings, and commercial performance.

- This project can be used to:

    - Predict movie success or revenue
    - Analyze trends in the film industry
    - Build recommendation systems
    - Study relationships between budget, popularity, and ratings
    - Perform clustering and classification of movie types

- Since the dataset contains some missing and inconsistent values, data preprocessing and cleaning are essential steps before analysis. Overall, this dataset provides a strong foundation for exploring real-world data science and machine learning techniques in the entertainment domain.

## Exploratory Data Analysis (EDA)

### 1. Understand the Dataset

#### 1.1 Import Libraries

In [1]:
import numpy as np
import pandas as pd


#### 1.2 Load the dataset

In [2]:
movies = pd.read_csv('dataset\\tmdb_5000_movies.csv')
credits = pd.read_csv('dataset\\tmdb_5000_credits.csv')

#### 1.3 Datasets Overview Before Merge

In [3]:
movies.head(1)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800


In [4]:
credits.head(1)

,movie_id,title,cast,crew
0,19995,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


#### 1.4 Merge Two DataFrames on the title Column
 - The following merge operation combines the movies and credits DataFrames using the common title column. This creates a unified dataset containing both movie details and cast/crew information for further analysis.And new dataset name is "movies".

In [5]:
movies = movies.merge(credits, on='title')

#### 1.5 Dataset Overview Using head() and tail()

In [6]:
movies.head(2)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,...,runtime,spoken_languages,status,tagline,title,vote_average,vote_count,movie_id,cast,crew
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",...,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800,19995,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...",...,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500,285,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."


In [7]:
movies.tail(2)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,...,runtime,spoken_languages,status,tagline,title,vote_average,vote_count,movie_id,cast,crew
4807,0,[],http://shanghaicalling.com/,126186,[],en,Shanghai Calling,When ambitious New York attorney Sam is sent t...,0.857008,[],...,98.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,A New Yorker in Shanghai,Shanghai Calling,5.7,7,126186,"[{""cast_id"": 3, ""character"": ""Sam"", ""credit_id...","[{""credit_id"": ""52fe4ad9c3a368484e16a36b"", ""de..."
4808,0,"[{""id"": 99, ""name"": ""Documentary""}]",NaN,25975,"[{""id"": 1523, ""name"": ""obsession""}, {""id"": 224...",en,My Date with Drew,Ever since the second grade when he first saw ...,1.929883,"[{""name"": ""rusty bear entertainment"", ""id"": 87...",...,90.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,NaN,My Date with Drew,6.3,16,25975,"[{""cast_id"": 3, ""character"": ""Herself"", ""credi...","[{""credit_id"": ""58ce021b9251415a390165d9"", ""de..."


#### 1.6 Checking Rows and Columns of the Dataset

In [8]:
movies.shape

(4809, 23)

- The dataset contains 4,809 rows and 23 columns, representing movie records and their related features.

#### 1.7 Understanding Each Feature/Column

In [9]:
movies.info()

<class 'pandas.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   str    
 2   homepage              1713 non-null   str    
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   str    
 5   original_language     4809 non-null   str    
 6   original_title        4809 non-null   str    
 7   overview              4806 non-null   str    
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   str    
 10  production_countries  4809 non-null   str    
 11  release_date          4808 non-null   str    
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   str    
 15  status                4809 non-n

- The dataset contains 4,821 movie records with 26 features, including numerical (budget, revenue, ratings) and categorical (title, genres, cast, crew) variables.
- It includes a mix of complete and missing values across certain columns like homepage, tagline, and overview.


#### 1.8 Statistical Summary

In [10]:
movies.describe()

,budget,id,popularity,revenue,runtime,vote_average,vote_count,movie_id
count,4.809000e+03,4809.000000,4809.000000,4.809000e+03,4807.000000,4809.000000,4809.000000,4809.000000
mean,2.902780e+07,57120.571429,21.491664,8.227511e+07,106.882255,6.092514,690.331670,57120.571429
std,4.070473e+07,88653.369849,31.803366,1.628379e+08,22.602535,1.193989,1234.187111,88653.369849
min,0.000000e+00,5.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,5.000000
25%,7.800000e+05,9012.000000,4.667230,0.000000e+00,94.000000,5.600000,54.000000,9012.000000
50%,1.500000e+07,14624.000000,12.921594,1.917000e+07,103.000000,6.200000,235.000000,14624.000000
75%,4.000000e+07,58595.000000,28.350529,9.291317e+07,118.000000,6.800000,737.000000,58595.000000
max,3.800000e+08,459488.000000,875.581305,2.787965e+09,338.000000,10.000000,13752.000000,459488.000000


- This table provides the statistical summary of numerical features such as budget, revenue, popularity, runtime, and ratings in the dataset.
- However, most of these features are not directly relevant for our content-based recommendation system, which mainly focuses on textual information like genres, cast, keywords, and overview.


### 2. Selecting Relevant Features for Recommendation System

In [11]:
movies = movies[['movie_id', 'title', 'overview', 'genres', 'keywords', 'cast', 'crew']]

#### 2.1 Dataset Overview After Selecting Relevant Features

In [12]:
movies.head(2)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...","[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...","[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...","[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."


#### 2.2 Overview of Important Features for Recommendation System

In [13]:
movies.info()

<class 'pandas.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   movie_id  4809 non-null   int64
 1   title     4809 non-null   str  
 2   overview  4806 non-null   str  
 3   genres    4809 non-null   str  
 4   keywords  4809 non-null   str  
 5   cast      4809 non-null   str  
 6   crew      4809 non-null   str  
dtypes: int64(1), str(6)
memory usage: 263.1 KB


- The dataset now contains 4,809 movie records with 7 selected features focused on recommendation system development.
- It includes a mix of numerical and text-based columns such as title, overview, genres, cast, and crew for building content-based filtering.


#### 2.3 Check Missing Values

In [14]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

- Only the **overview** column has 3 missing values, while all other columns have complete data with no null entries.

#### 2.4 Remove Unnecessary Null Values

In [15]:
movies.dropna(inplace=True)

- Removes all rows containing missing (NaN) values from the dataset and updates the DataFrame in place, ensuring a clean dataset for further processing.

#### 2.5 Check Duplicate Records

In [16]:
print(movies.duplicated().sum())

0


- The dataset contains no duplicate rows, ensuring all records are unique.

### 3. Feature Engineering from Categorical Data

#### 3.1 Preparing Genres for Model Building
- We analyze the genres column as it is a key categorical feature for the recommendation system.
- Since it is stored in JSON-like string format, we first inspect its structure to properly extract useful genre information.


In [17]:
import ast

def convert(obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

movies['genres'] = movies['genres'].apply(convert)

In [18]:
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


#### 3.2 Preparing keywords for Model Building

In [19]:
movies['keywords'] = movies['keywords'].apply(convert)
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


#### 3.3 Preparing cast for Model Building

In [20]:
movies['cast'] = movies['cast'].apply(convert)
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[Sam Worthington, Zoe Saldana, Sigourney Weave...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


In [21]:
def convert3(obj):
    L = []
    counter = 0
    
    for i in obj:
        if counter != 3:
            L.append(i)
            counter += 1
        else:
            break         
    return L

movies['cast'] = movies['cast'].apply(convert3)
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[Sam Worthington, Zoe Saldana, Sigourney Weaver]","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."


- Above ouput shows that only the first three cast members from each movie’s cast list.
- Limiting the cast to top actors helps reduce unnecessary data and keeps the recommendation system more focused and efficient.


#### 3.4 Preparing crew for Model Building

In [22]:
def fetch_director(text):
    L = []
    for i in ast.literal_eval(text):
        if i['job'] == 'Director':
            L.append(i['name'])
    return L 

movies['crew'] = movies['crew'].apply(fetch_director)

In [23]:

movies.head(1)  

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, Science Fiction]","[culture clash, future, space war, space colon...","[Sam Worthington, Zoe Saldana, Sigourney Weaver]",[James Cameron]


#### 3.5 Removing Spaces from Multi-word Features

In [24]:
def collapse(L):
    L1 = []
    for i in L:
        L1.append(i.replace(" ",""))
    return L1

In [25]:
movies['cast'] = movies['cast'].apply(collapse)
movies['crew'] = movies['crew'].apply(collapse)
movies['genres'] = movies['genres'].apply(collapse)
movies['keywords'] = movies['keywords'].apply(collapse)

In [26]:
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[Action, Adventure, Fantasy, ScienceFiction]","[cultureclash, future, spacewar, spacecolony, ...","[SamWorthington, ZoeSaldana, SigourneyWeaver]",[JamesCameron]


#### 3.6 Transforming Overview Sentences into Word Lists

In [27]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())

In [28]:
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin...","[Action, Adventure, Fantasy, ScienceFiction]","[cultureclash, future, spacewar, spacecolony, ...","[SamWorthington, ZoeSaldana, SigourneyWeaver]",[JamesCameron]


- This show that converts each movie overview from a sentence into a list of individual words (tokens).
- Tokenizing the text helps the recommendation system analyze and compare movie descriptions more effectively.


#### 3.7 Combining All Text Features into a Single Tags Column

In [29]:
movies["tags"] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [30]:
movies.head(1)

,movie_id,title,overview,genres,keywords,cast,crew,tags
0,19995,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin...","[Action, Adventure, Fantasy, ScienceFiction]","[cultureclash, future, spacewar, spacecolony, ...","[SamWorthington, ZoeSaldana, SigourneyWeaver]",[JamesCameron],"[In, the, 22nd, century,, a, paraplegic, Marin..."


- This show that creates a new feature called tags by combining important textual features like overview, genres, keywords, cast, and crew into a single column.
- This helps the recommendation system capture all relevant movie information in one place for better similarity matching.

#### 3.8 Final Dataset for Recommendation System

In [31]:
new_df = movies[['movie_id', 'title', 'tags']]
new_df.head(1)

,movie_id,title,tags
0,19995,Avatar,"[In, the, 22nd, century,, a, paraplegic, Marin..."


- This creates a new DataFrame by selecting only the essential columns: movie_id, title, and tags.
- It helps reduce unnecessary data and prepares a clean dataset focused only on information required for building the recommendation system.

In [32]:
new_df['tags'] = new_df['tags'].apply(lambda x: " ".join(x))
new_df.head(3)


,movie_id,title,tags
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha..."
2,206647,Spectre,A cryptic message from Bond’s past sends him o...


- This converts the tags column from a list of words into a single space-separated string.
- It is done to make the data suitable for text vectorization in the recommendation system.

#### 3.9 Converting Tags Text to Lowercase

In [33]:
new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())

In [34]:
new_df.head()

,movie_id,title,tags
0,19995,Avatar,"in the 22nd century, a paraplegic marine is di..."
1,285,Pirates of the Caribbean: At World's End,"captain barbossa, long believed to be dead, ha..."
2,206647,Spectre,a cryptic message from bond’s past sends him o...
3,49026,The Dark Knight Rises,following the death of district attorney harve...
4,49529,John Carter,"john carter is a war-weary, former military ca..."


#### 3.10 Text Normalization using Porter Stemmer

In [35]:

import nltk
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()


- This code imports the NLTK library and the PorterStemmer, which is used to reduce words to their root form (e.g., loved → love, running → run).
- It helps improve text processing in the recommendation system by treating similar words as the same token.

In [36]:
def stem(text):
    y = []
    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

new_df['tags'] = new_df['tags'].apply(stem)

In [37]:
new_df.head(1)

,movie_id,title,tags
0,19995,Avatar,"in the 22nd century, a parapleg marin is dispa..."


### 4. Feature Extraction (Model Preparation Stage)

#### 4.1 Converting Text into Numerical Features Using CountVectorizer

In [38]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000, stop_words='english')

- This code initializes CountVectorizer from Scikit-learn to convert text data into numerical form by creating a bag-of-words representation.
- We limit the features to the top 5000 most frequent words and remove English stop words(is, am, are) to improve model efficiency and reduce noise.

In [39]:
vectors = cv.fit_transform(new_df['tags']).toarray()

In [40]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0], shape=(5000,))

#### 4.2 Building Similarity Matrix for Recommendation System

In [41]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vectors)

- This code computes the cosine similarity matrix between all movie vectors.
- It measures how similar two movies are based on their text features, which helps in recommending movies with similar content.

In [42]:
similarity.shape

(4806, 4806)

In [43]:
similarity[0]

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ], shape=(4806,))

#### 4.3 Movie Recommendation Function Using Similarity Scores

In [44]:
def recommend(movie):
    movie_index = new_df[new_df['title'] == movie].index[0]
    distances = similarity[movie_index]
    movies_list = sorted(list(enumerate(distances)), reverse=True, key=lambda x: x[1])[1:6]
    
    for i in movies_list:
        print(new_df.iloc[i[0]].title)

- This function recommends movies based on a given movie title by finding its index, retrieving similarity scores from the cosine similarity matrix, and sorting them in descending order.
- It then prints the top 5 most similar movies (excluding the input movie itself) from the dataset.

In [45]:
recommend('Gandhi')

Gandhi, My Father
Guiana 1838
The Wind That Shakes the Barley
Mr. Turner
A Passage to India


#### 4.4 Saving Processed Data for Deployment

In [46]:
import pickle
# pickle.dump(new_df, open('movies.pkl', 'wb'))

In [47]:
pickle.dump(new_df,open('movie_list.pkl','wb'))
pickle.dump(similarity,open('similarity.pkl','wb'))

In [48]:
# new_df['title'].values

In [49]:
# pickle.dump(new_df.to_dict(), open('movies_dict.pkl', 'wb'))

In [50]:
# pickle.dump(similarity, open('similarity.pkl', 'wb'))